### ISSUE TITLE
Having trouble binding to null.String in a PostgreSQL DB with nullable columns

### ISSUE THREAD

**Issue by tanishq-dubey** [volatiletech/sqlboiler#456]
If you're having a generation problem please answer these questions before submitting your issue. Thanks!

### What version of SQLBoiler are you using (`sqlboiler --version`)?
SQLBoiler v3.1.0

### If this happened at runtime what code produced the issue? (if not applicable leave blank)
This is the relevant code that would produce the error:

```
import (
    "github.com/volatiletech/null"
    "github.com/volatiletech/sqlboiler/queries"
)
type MyQueryData struct {
	FsymID      string `boil:"fsym_id" json:"fsym_id" toml:"fsym_id" yaml:"fsym_id"`
	Name        *string `boil:"name" json:"name" toml:"name" yaml:"name"`
}

db, err := sql.Open("postgres", config.GetString("postgres.conn"))
var queryData MyQueryData
if err := queries.Raw("SELECT distinct myDBSchema.coverage.fsym_id, myDBSchema.bbg.bbg_ticker as name FROM myDBSchema.coverage LEFT JOIN myDBSchema.bbg ON myDBSchema.coverage.fsym_id=myDBSchema.bbg.fsym_id WHERE myDBSchema.sym_v1_sym_coverage.fsym_id = $1;", "SOMESTRINGID").Bind(context.Background(), db, &queryData);  err != nil {
    panic(err)
}
fmt.Println(queryData) // the `name` field is blank here (valid = false)
```

### Please provide a relevant database schema so we can replicate your issue (Provided you are comfortable sharing this)
These are the two tables I reference in my query
```
create table myDBSchema.bbg
(
  fsym_id    char(8)  not null
     constraint bbg_pkey
      primary key,
  bbg_id     char(12) not null,
  bbg_ticker varchar(20)
);
```

```
create table myDBSchema.coverage
(
  fsym_id                 char(8)      not null
    constraint coverage_pkey
      primary key,
  currency                char(3),
  proper_name             varchar(200) not null,
  primary_equity_id  char(8),
  primary_listing_id char(8),
  active_flag             smallint     not null,
  security_type        varchar(40),
  listing_exchange   varchar(4),
  listing_flag            smallint     not null,
  regional_flag           smallint     not null,
  security_flag           smallint     not null,
  regional_id        char(8),
  security_id        char(8)      not null,
  universe_type           char(2)      not null
);
```

### Further information. What did you do, what did you expect?
I am running a raw query to grab some data by joining two tables on a **Postgres 9.6** database. I have detailed which fields can be null in the schema document above. I am simply doing a `JOIN` on the two tables, and then getting the  result, the SQL query can be seen here:
```
SELECT distinct myDBSchema.coverage.fsym_id, 
    myDBSchema.bbg.bbg_ticker as name 

FROM myDBSchema.coverage 
LEFT JOIN myDBSchema.bbg ON myDBSchema.coverage.fsym_id=myDBSchema.bbg.fsym_id 
WHERE myDBSchema.coverage.fsym_id = 'SOMESTRINGID';
```

Running this query in other database tools such as DataGrip, PSQL, etc returns the columns I am expecting: a FSYM_ID and NAME,  both as strings. 

When I run this same query with SQLBoiler, as detailed with the code above, I do not get an error, but my name field is marked as non valid (`name.valid == false`).

I have tried a couple things, specifically changing the type of the name to `string` and `*string` but SQLBoiler correctly fails to bind to my struct, as the types don't match. As I was debugging this, I also noticed that upon inspecting
[BindMapping() ](https://github.com/volatiletech/sqlboiler/blob/master/queries/reflect.go#L301), the mapping array here has valid pointers for both my ID and name.

It then seems like SQLBoiler cannot map a row with all values populates (ID and Name) to a `null.String` type. What could be the cause of this/workaround?

**Comment by aarondl** [comment]
Hey there. Unfortunately `left join` isn't really a supported use case. That sounds insane, and it kind of is.

The problem is this:
The resulting row returned by a database for something that is being left joined can look like the following

```text
id | name | id   | name
-----------------------
1  | 'hi' | null | null
```

The problem with this is that `Bind` works by putting values directly into your structs. So typically when doing an inner join with a raw query you're required to create a struct that can bind both parts of the returned row.

```go
type myInnerJoinReturnStruct struct {
    A models.A
    B models.B
}
```

So `Bind` will take a row and by having mangled the names of the query to actually be `A.id` and `B.id` so that it can actually tell what columns belong to which part of that return struct and slap the values into the correct spots.

So there's two limitations that could possibly be fixed:
1. Bind does not deal with `new`ing any structs so your structs in the return struct cannot be pointers. This means in the land of `left join` we don't have a current mechanism to support leaving one model nil and the other actually be something. This is admittedly simple to fix (just a bit more reflect than we have now) as long as the hard part could be taken care of (detecting a null row).
2. Bind currently takes a row value and directly inserts it into a field value it finds by following the Bind Mapping. This means that there's no provision for suddenly the entire row being null as it will try to bind null to a `string` field value and fail (pretty sure this is the problem you're running into).

The only solution I have in my head is to (maybe conditionally in the case of joins) bind not directly to the struct fields - but to an []interface{} whose length matches the number of fields of possibly null struct (or maybe simply the entire row). We then check if all the interface values are nil, if they are we set the pointer in the join struct to nil. If they are not we begin to imitate what `Scan` does, going through those interface{} values and doing conversions and field setting using the `Scanner` interface. Though that's under the assumption `Scan` cannot be called twice for a given *Rows object, which maybe it can! (I should look into this, because that simplifies a lot).

So why if there's a solution hasn't it been done? Well it's an ugly solution, I was sort of hoping someone else would come along with a better one. It complicates the current code drastically and because of the intermediate layer of []interface{} binding ends up consuming more memory and slowing down every normal query just to support outer/left/right join.

There's no good answers or workarounds here unfortunately. You simply can't use the sqlboiler generated structs for left joins (even if you did it yourself) because not all field values can be nil. So you'd have to do the "is this entire struct going to end up being a null?" check yourself by doing that intermediate Scan into an []interface{}. After that though there's not even a pleasant way to use `Bind` to put things into the models since the way `Bind` works you'd have to have 3 different return structs.

Sorry to give you nothing but bad news. I'd be very happy to see other solutions to this problem. Maybe there's something less bad than what I'm describing here.

**Comment by tanishq-dubey** [comment]
Thanks for the prompt response, this was really helpful!

So I have a 2 fold response:

1. I was able to fix my issue due to some corruption in my vendor directory (doing a `dep ensure` and a clean build overall fixed things, and queries work as expected now)

2. I think you can support `LEFT JOIN` but with very carefully crafted queries. Specifically, if your query is careful to not include duplicate names in the response you _should be_™️ okay. For example:
```
SELECT table1.name as specialname, table2.name as morespecialname, ...
```

Here you are selecting the column `name` from these tables, but your query renames them to unique names, so the response will have unique names that you can bind to. This isn't really a SQLBoiler solution, but it does allow for the use of SQLBoiler to manage database queries (AKA, avoiding writing `db.open()`, `db.close()`...). In addition to this, your points still apply: I would still need to create my own structs with proper naming, etc.

So in conclusion, I did get `LEFT JOIN` working, but with these rules:
 - You must have each returned column have a unique name in your query (as shown with the above SQL statement)
 - You must write your own structs

With those rules, you can do nullable and whatever else. Perhaps we can list this in the documentation, but I would be hesitant on that, because I am not sure how universal these rules are. I ran my query more than 5K times without failure, so I know that works, but SQL is a big world.

Thanks again!

**Comment by tanishq-dubey** [comment]
I'll post a full sample (or even push a project to my github) later tonight! Hopefully that should help!
